### ISSUE TITLE
Proposed enhancements

### ISSUE THREAD

**Issue by anxiousmodernman** [bradfordboyle/fitbit-grabber-rs#1]
Nice cli tool. I was looking for something like this

Some ideas for enhancements:
1. Provide an additional config path for config under the standard-ish `$HOME/.config/fitbit-grabber`'
    * we can look for .token in the current directory first and then fall back to a user-level config dir
    * we can provide a config file as a fall back if env vars are unset
    * we can use this dedicated directory for cached data
2. Change the crate to function as both a library and cli binary
    * There's no wrapper for the fitbit API on crates.io yet, [as far as I can tell](https://crates.io/search?q=fitbit), so you could save people a few hundred lines of code
3. Cache/persist data
    * I'd like to run this on a timer and accumulate the data
    * we could provide summaries and views with rust, and/or provide a way to export something parsable
4. Additional data categories
    * Of immediate interest to me is the [body category](https://dev.fitbit.com/build/reference/web-api/body/) for weight, BMI, etc.

I'll probably hack 1 and 4 first for my own usage. Would you be interested in PRs with enhancements like these?

**Comment by bradfordboyle** [comment]
I just saw this now---not sure why I didn't get an email notification. Your proposals all sound great, please feel free to send PRs when you have them.

**Comment by anxiousmodernman** [comment]
Some more details about what I'm working toward. The API is big, and given the separate concerns of a cli and an API client, I will first scaffold out the end-to-end concerns of the cli (fetching, storing, viewing data) with a subset of the API first (body weight + activities). Over time, if we fill out the rest of the API, you can put it into a separate crate.

### Project structure
```
/src
  /bin
    main.rs   # binary "fitbit-grabber"; cli code
    config.rs # cli config
  lib.rs    # library "fitbit"
  errors.rs  # library errors (wrap things that can go wrong in lib)
```
### Library API
The core struct is still `FitbitClient`, and we can implement the big resource categories of the Fitbit API as traits: `Body`, `Activity`, `Food`, etc. Library users import the client and the trait implementations they need, so the namespace isn't as cluttered.  Example (the only one I got right now): 

![fitbit](https://user-images.githubusercontent.com/396004/40997727-81895704-68ba-11e8-8ac6-8dbaf6745103.png)

Result types can be container structs containing arrays, from json like
```
{ "weight": [ {measurement}, {measurement}, ...] }
```
We want to re-serialize the inner array for persistent storage.

### Persistence 

Persistence will be a concern of cli code under /bin. We can use a file-based key-value store, sled. For each result type we care to persist, we devise a deterministic key serialization strategy based on the fields, preferably some sort of date or datetime. If one of the `measurement`s above looked like the following

```
# From "weight" for Body time series with a period of 1 day
{ "date": "2016-12-19", "value": "85.0" }
```
We could derive a sortable key based on naively converting date into a full `chrono::DateTime` and rendering it, after some metadata prefix

```
# key -> value
/body/time-series/weight/1d/2016-12-19T00:00:00-08:00 -> { date: d, value: x }
     derived from query <- ^ -> derived from data       
```
Given similar query periodicity parameters (which we _could_ store in the key, as shown), it should be possible just re-fetch in whole or in part without corrupting the db, since the primary key is derived from the data itself, plus stuff we know about the query at runtime.

*Timezone*, required to convert a `NaiveDate` into an RFC3339-renderable `DateTime`, is a changeable _global_ property under the user profile in Fitbit's API. We need to discover the tz offset from the profile first, because Fitbit times are always relative to this value.

It would be *nice* to directly deserialize the strings like "2016-12-19" and "85.0" into `NaiveDate` and `f64` respectively, but I don't know how to do that in serde, yet.

### Views
It would be good to drive all views from the database. E.g. serve cached data first.

**Comment by bradfordboyle** [comment]
I had started working along a similar line of thinking w.r.t splitting into a binary and library but I never got around to pushing. See fc9bd23405497142337f7f5076a7366eb7e75733 for an example of where I was going.

I have not thought much about the persistence side of things. My "workflow" was to grab the raw data as JSON and then use something like python/clojure for post-processing to generate things like graphs and or tables (e.g., average daily heart rate over the past six months) or correlate with other data sources (e.g., daily blood pressure readings). Is adding persistence combining two different concerns into one tool? I guess what I'm trying to understand is what are the pros and cons of

```sh
# w/ persistence
fitbit-grabber step --date 2016-07-12

# pipe the output into a separate utility that handles persistence
fitbit-grabber step --date 2016-07-12 | fitbit-db
```

That being said, I think there may be use-case for adding caching. Looking at the example key -> value mapping, the key almost looks like the URL path and I wonder if we should just think of persistence as caching URL responses? A lot of this data is historical and should never change once it is recorded. If we cache a URL request and it's response, then subsequent runs of the CLI can query the cache and avoid API calls. We could add flags for invalidating the cache (clear all entries) and for skipping the cache for this call.

**Comment by anxiousmodernman** [comment]
Yeah, I don't want to break your "pipe to script" use case. That's a nice thing to have. There's a good chance that we're developing for different use cases at the application level.

I guess the place where our concerns really overlap is the api library. I think I'll open a PR just with those changes, and very little else on the cli side. 

In the meantime, I could develop my own database-y app in another repo, and add your library as a dependency with cargo's `git` or `path`.

**Comment by anxiousmodernman** [comment]
One reason I wanted to reach for a db was to store dates and timestamps in a strongly-typed, information-preserving way. As I mentioned earlier, fitbit is returning strings for the body weight time series. While that kind of annoys me on a personal level, it's probably better to not fight it in the library, and model the api reponses exactly as they are.

I was actually getting hung up in my own app because I wanted to convert `String` -> `chrono::DateTime<Tz>`. But that doesn't have to be a blocker, strictly speaking.

If they API is modeled as structs, however, you will need to add a couple lines of code to re-serialize the types as JSON. But `serde` will let you do derive that automatically for ordinary structs.
